onnxscript/type_annotation.py

```diff
@@ -28,31 +28,91 @@
 _LIST_CONSTRUCTORS = frozenset([list, typing.List, typing.Sequence, collections.abc.Sequence])
 
 
-def pytype_to_attrtype(pytype: type) -> typing.Optional[onnx.AttributeProto.AttributeType]:
+def remove_annotation(typeinfo):
+    """Remove Annotated wrapper if present, otherwise return typeinfo as is."""
+    if hasattr(typing, "Annotated"):
+        # Present in Python 3.9+
+        if get_origin(typeinfo) is typing.Annotated:
+            return get_args(typeinfo)[0]
+    return typeinfo
+
+
+def is_primitive_attr_type(typeinfo) -> bool:
+    return typeinfo in _PYTYPE_TO_ATTRTYPE_MAP
+
+
+def pytype_to_attrtype(pytype) -> typing.Optional[onnx.AttributeProto.AttributeType]:
+    pytype = remove_annotation(pytype)
     if pytype in _PYTYPE_TO_ATTRTYPE_MAP:
         return _PYTYPE_TO_ATTRTYPE_MAP[pytype]
-    if get_origin(pytype) in _LIST_CONSTRUCTORS:
-        args = get_args(pytype)
-        elt_type = args[0]
+    type_constructor = get_origin(pytype)
+    # Remove Optional wrapper if present, which is represented as an Union[..., None]
+    if type_constructor is typing.Union:
+        # Filter out None, since typing.Optional[X] evaluates to Union[X, None]
+        args = [x for x in get_args(pytype) if x is not type(None)]  # noqa: E721
```

Comment:
Just making sure, is type(None) None?

Reply:
No. I believe the easiest way to specify the type of None is `type(None)`.

Reply:
I see. Just thought `Union[x, None]` will contain None but I must be wrong

Reply:
Oh, I think it's my comment that is misleading ... I should say `Union[X, type(None)]` in the comment ... I think the typing support implicitly treats (promotes) `None` as `type(None)` but they are distinct values.